File: Python/utils/csv_reader.py
from typing import List
from datetime import datetime
import csv

from models.note import Note


CSV_FILE = "Problem/input_data.csv"


class CsvReader:
    def __init__(self):
        self._notes: List[Note] = []

    @property
    def notes(self) -> List[Note]:
        return self._notes

    def _read_data(self, csv_file: str) -> List[Note]:
        with open(csv_file, "r", newline='') as f:
            reader = csv.reader(f)
            next(reader)  # skip the header
            result = []
            for row in reader:
                data = Note(
                    row[0].strip(),
                    row[1].strip(),
                    row[2].strip().split(" ")
                    if row[2].strip()
                    else [],
                    datetime.strptime(
                        row[3].strip(),
                        "%m/%d/%Y"),
                    datetime.strptime(
                        row[4].strip(),
                        "%m/%d/%Y"),
                )
                result.append(data)
            return result

    def read_notes(self) -> None:
        self._notes = self._read_data(CSV_FILE)

File: Go/utils/csv_reader.go
package utils

import (
	"encoding/csv"
	"fmt"
	"os"
	"strings"

	"github.com/siddhant-vij/Note-Taking-App/models"
)

const CSV_FILE = "Problem/input_data.csv"

type CSVReader struct {
	notes []models.Note
}

func (cr *CSVReader) Notes() []models.Note {
	return cr.notes
}

func (cr *CSVReader) readData(csvFile string) ([]models.Note, error) {
	file, err := os.Open(csvFile)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	if _, err := reader.Read(); err != nil {
		return nil, err
	}

	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	result := []models.Note{}
	for _, row := range records {
		tags := []string{}
		if strings.TrimSpace(row[2]) != "" {
			tags = strings.Split(row[2], " ")
		}
		createdOn, err := models.ParseDate(strings.TrimSpace(row[3]))
		if err != nil {
			return nil, fmt.Errorf("error parsing CreatedOn date: %v", err)
		}
		lastUpdatedOn, err := models.ParseDate(strings.TrimSpace(row[4]))
		if err != nil {
			return nil, fmt.Errorf("error parsing LastUpdatedOn date: %v", err)
		}
		
		result = append(result, models.Note{
			Title:         strings.TrimSpace(row[0]),
			Content:       strings.TrimSpace(row[1]),
			Tags:          tags,
			CreatedOn:     createdOn,
			LastUpdatedOn: lastUpdatedOn,
		})
	}

	return result, nil
}

func (cr *CSVReader) ReadNotes() error {
	notes, err := cr.readData(CSV_FILE)
	if err != nil {
		return err
	}
	cr.notes = notes
	return nil
}
